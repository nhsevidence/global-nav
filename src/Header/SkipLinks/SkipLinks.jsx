import React, { Component } from "react";
import PropTypes from "prop-types";

import styles from "./SkipLinks.module.scss";

export default class SkipLinks extends Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);
	}

	handleClick(e) {
		const href = e.currentTarget.getAttribute("href");

		if (href && href.indexOf("#") === 0) {
			const id = href.split("#")[1],
				element = document.getElementById(id);

			if (element) {
				e.preventDefault();

				element.setAttribute("tabIndex", "-1");
				element.focus();

				window.scrollTo(0, element.offsetTop);
			} else {
				console.warn(`Skip link target with id ${id} doesn't exist`);
			}
		}
	}

	render() {
		return (
			<ul className={styles.list} aria-label="Accessibility links">
				<li>
					<a
						href={`#${this.props.skipLinkId}`}
						className={styles.link}
						onClick={this.handleClick}
					>
						Skip to content
					</a>
				</li>
				<li>
					<a
						href="https://www.nice.org.uk/accessibility"
						className={styles.link}
					>
						Accessibility help
					</a>
				</li>
			</ul>
		);
	}
}

SkipLinks.propTypes = {
	skipLinkId: PropTypes.string.isRequired
};
