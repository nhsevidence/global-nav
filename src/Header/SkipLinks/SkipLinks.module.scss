@import '~@nice-digital/design-system/src/stylesheets/settings/settings';
@import '~@nice-digital/design-system/src/stylesheets/helpers/helpers-utils';

.list {
  list-style: none;
  margin: 0;
  padding: 0;

  li {
    list-style: none;
    margin: 0;
  }
}

.link {
  clip: rect(1px 1px 1px 1px);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;

  &:focus {
    background: #fff;
    clip: auto;
    color: $colour-link;
    display: block;
    height: auto;
    left: rem($spacing-x-small);
    outline: 3px solid $colour-focus;
    overflow: visible;
    padding: rem($spacing-small $spacing-medium);
    position: absolute;
    top: rem($spacing-x-small);
    white-space: nowrap;
    width: auto;
    z-index: 1;
  }
}
