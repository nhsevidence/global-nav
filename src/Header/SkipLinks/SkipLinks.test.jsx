import React from "react";
import SkipLinks from "./SkipLinks";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

describe("SkipLinks", () => {
	const skipLinkId = "test-id";
	let contentDiv;

	beforeEach(() => {
		// Create the skip target
		contentDiv = document.createElement("div");
		contentDiv.id = skipLinkId;
		document.body.append(contentDiv);

		// Fake window.scrollTo because it isn't implemented in jsdom AND we want to spy on it
		window.scrollTo = jest.fn();
	});

	afterEach(() => {
		Object.defineProperties(window.HTMLElement.prototype, {
			offsetTop: {
				get: () => 0
			}
		});

		contentDiv.parentNode.removeChild(contentDiv);
	});

	it("Renders without crashing", () => {
		const wrapper = shallow(<SkipLinks skipLinkId="test" />);
		expect(wrapper).toHaveLength(1);
	});

	it("Matches snapshot", () => {
		const wrapper = shallow(<SkipLinks skipLinkId="test" />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Renders skip link id href from prop", () => {
		const wrapper = shallow(<SkipLinks skipLinkId="test" />);
		expect(
			wrapper
				.find("a")
				.at(0)
				.props().href
		).toEqual("#test");
	});

	it("Triggers click handler on skip to content click", () => {
		const handleClick = jest.spyOn(SkipLinks.prototype, "handleClick");
		const wrapper = shallow(<SkipLinks skipLinkId="test" />);

		wrapper
			.find("a")
			.at(0)
			.simulate("click", {
				currentTarget: {
					getAttribute: () => "#" + skipLinkId
				},
				preventDefault: () => {}
			});

		expect(handleClick).toHaveBeenCalledTimes(1);
	});

	it("Prevents default and moves focus to the skip link target on click", () => {
		const wrapper = shallow(<SkipLinks skipLinkId="test" />);

		const preventDefault = jest.fn();

		wrapper.instance().getYOffset = jest.fn(() => 99);

		wrapper
			.find("a")
			.at(0)
			.simulate("click", {
				currentTarget: {
					getAttribute: () => "#" + skipLinkId
				},
				preventDefault: preventDefault
			});

		expect(preventDefault).toHaveBeenCalledTimes(1);
		expect(contentDiv.getAttribute("tabIndex")).toEqual("-1");
		expect(document.activeElement).toBe(contentDiv);
		expect(window.scrollTo).toHaveBeenCalledWith(0, 99);
	});

	describe("getYOffset", () => {
		it("should get offset top relative to window", () => {
			const wrapper = shallow(<SkipLinks skipLinkId="test" />);

			var element = {
				offsetLeft: 0,
				offsetTop: 99,
				scrollTop: 9,
				offsetParent: {
					offsetLeft: 0,
					offsetTop: 98,
					scrollTop: 8
				}
			};

			expect(wrapper.instance().getYOffset(element)).toBe(180);
		});
	});
});
