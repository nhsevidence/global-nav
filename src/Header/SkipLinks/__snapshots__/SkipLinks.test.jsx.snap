// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkipLinks Matches snapshot 1`] = `
<ul
  aria-label="Accessibility links"
  className="list"
>
  <li>
    <a
      className="link"
      href="#test"
      onClick={[Function]}
    >
      Skip to content
    </a>
  </li>
  <li>
    <a
      className="link"
      href="https://www.nice.org.uk/accessibility"
    >
      Accessibility help
    </a>
  </li>
</ul>
`;
